### ISSUE TITLE
GEO Location data is truncated Hyundai USA

### ISSUE THREAD

**Issue by rappazzo** [fuatakgun/kia_uvo#196]
Please check Services, Known Bug / Issues and Troubleshooting over here first: https://github.com/fuatakgun/kia_uvo/blob/master/README.md
**Region and Brand of car**
USA - Hyundai 2020 Ioniq EV

**Describe the bug**
The Location info in the logs seems to have lost its decimal precision
Here is my car's current location ([in the Atlantic Ocean](https://www.latlong.net/c/?lat=40&long=-74)!):
```
         'vehicleLocation': {
            'coord': {
               'alt': 24, 
               'lon': -74, 
               'type': 0, 
               'lat': 40
            }
         }, 
 ```
It looks like this is just truncated (not rounded), as my actual latitude is closer to 41

**To Reproduce**
1. In the lovelace interface, add a map card including the device tracker location of the car
2. Observe the imprecise location.
3. In the logs, observe that the location lat and lon values have no decimal precision

**Expected behavior**
The lat and lon values need a few more decimal places for more accuracy

**Screenshots**
<img width="1012" alt="From Home Assistant" src="https://user-images.githubusercontent.com/525287/144311083-042618b1-aa7f-4698-89c2-f91afb2bbcb3.png">
("mep" is my phone's location)

**Comment by cdnninja** [comment]
Are you able to confirm the Kia app works correctly?

**Comment by rappazzo** [comment]
yeah here is a screen shot (Hyundai app):
![From Hyundai App](https://user-images.githubusercontent.com/525287/144313030-3b503e2e-293d-4ca6-99cb-c8c92afcf8ab.png)
:

**Comment by cdnninja** [comment]
Is the above from the logs or home assistant data? I don't see us doing anything to the location so not sure why this would occur unless they aren't giving us the right amount of accuracy.

**Comment by rappazzo** [comment]
The above data is from the logs (after I turned on debug level logging).  For what it's worth, I used to use Bluelinky with Node Red, and that did have the correct precision in the data.

**Comment by rappazzo** [comment]
Here is the full message from the log: 
```
2021-12-01 15:06:44 DEBUG (SyncWorker_4) [custom_components.kia_uvo.HyundaiBlueLinkAPIUSA] kia_uvo - get_cached_vehicle_status response {'hataTID': '<redacted>', 'vehicleStatus': {'dateTime': '2021-12-01T20:05:12Z', 'acc': False, 'defrostStatus': 'false', 'transCond': True, 'doorLockStatus': 'false', 'doorOpen': {'frontRight': 0, 'frontLeft': 0, 'backLeft': 0, 'backRight': 0}, 'battery': {'batSoc': 87, 'batState': 0, 'sjbDeliveryMode': 0}, 'vehicleLocation': {'coord': {'alt': 24, 'lon': -74, 'type': 0, 'lat': 40}}, 'ign3': False, 'ignitionStatus': 'false', 'lowFuelLight': False, 'sideBackWindowHeat': 0, 'engine': False, 'hoodOpen': False, 'airConditionStatus': 'false', 'steerWheelHeat': 0, 'trunkOpen': False, 'doorLock': False, 'airCtrlOn': False, 'airTemp': {'unit': 1, 'hvacTempType': 0, 'value': 'LO'}, 'evStatus': {'valueDiff': -2, 'remainTime2': {'etc3': {'unit': 1, 'value': 25}, 'etc2': {'unit': 1, 'value': 0}, 'atc': {'unit': 1, 'value': 0}, 'etc1': {'unit': 1, 'value': 0}}, 'evIgnitionStatus': False, 'batteryPlugin': 0, 'timeDiff': 35703, 'batteryCharge': False, 'batteryStatus': 98, 'remainTime': [], 'drvDistance': [{'rangeByFuel': {'gasModeRange': {'unit': 3, 'value': 0.0}, 'totalAvailableRange': {'unit': 3, 'value': 200.0}, 'evModeRange': {'unit': 3, 'value': 200.0}}, 'type': 2}], 'reservChargeInfos': {'targetSOClist': [{'rangeByFuel': {'gasModeRange': {'unit': 3, 'value': 0}, 'totalAvailableRange': {'unit': 3, 'value': 200}, 'evModeRange': {'unit': 3, 'value': 200}}, 'plugType': 0, 'targetSOClevel': 127, 'type': 2}, {'rangeByFuel': {'gasModeRange': {'unit': 3, 'value': 0}, 'totalAvailableRange': {'unit': 3, 'value': 200}, 'evModeRange': {'unit': 3, 'value': 200}}, 'plugType': 1, 'targetSOClevel': 127, 'type': 2}], 'reservChargeInfo': {'dateTime': '2021/12/01T12:06:44Z', 'reservChargeInfoDetail': {'reservChargeSet': False, 'reservEndTime': '', 'reservFatcSet': {'defrost': False, 'airTemp': {'unit': 0, 'value': '00H'}, 'airCtrl': 0}, 'reservInfo': {'time': {'timeSection': 0, 'time': '1200'}, 'day': [9]}}}, 'reservFlag': 0, 'reserveChargeInfo2': {'reservChargeInfoDetail': {'reservChargeSet': False, 'reservEndTime': '', 'reservFatcSet': {'defrost': False, 'airTemp': {'unit': 0, 'value': '00H'}, 'airCtrl': 0}, 'reservInfo': {'time': {'timeSection': 0, 'time': '1200'}, 'day': [9]}}}, 'offpeakPowerInfo': {'offPeakPowerFlag': 0, 'offPeakPowerTime1': {'startTime': {'timeSection': 0, 'time': '1200'}, 'endTime': {'timeSection': 0, 'time': '1200'}}}}}, 'sleepModeCheck': True, 'defrost': False, 'tirePressureLamp': {'tirePressureWarningLampRearLeft': 0, 'tirePressureWarningLampFrontLeft': 0, 'tirePressureWarningLampFrontRight': 0, 'tirePressureWarningLampAll': 0, 'tirePressureWarningLampRearRight': 0}, 'trunkOpenStatus': 'false'}}```

**Comment by rappazzo** [comment]
I am primarily a java coder so take this with a grain of salt, but it looks like `custom_components/kia_uvo/HyundaiBlueLinkAPIUSA.py` doesn't implement a separate call to get the location.  Having looked through the bluelinky node red code (it's in node), there is a separate call for location and odometer.  Not sure if that is relevant or not

**Comment by cdnninja** [comment]
get_location was used in the canada API because the data wasn't included in the first API call.  It appears as though Hyundai USA includes atleast some data in the initial call.  

I think you are correct in that we need to call a second API to get the more accurate data.  

Today we don't have a developer actively working on Hyundai USA API, I cover canada and other regions where I can.  Feel free to drop a PR if you are inclined.   I can also give it the first go around if you would rather not, I would need a hand testing though.

**Comment by rappazzo** [comment]
I could take it up, but I don't know how to setup the testing environment.  Can you point me to how to do that?

**Comment by cdnninja** [comment]
I am terrible and just edit my production HA code, restarting HA all the time.  I have it so I can browse my HA via SMB and edit via my editor of choice. 

  @fuatakgun  recommended I setup a dev environment and had thoughts on a way that works well.

**Comment by fuatakgun** [comment]
i have setup vs code following home assistant developer guidelines which us independent than my production environment. Restarting it is very fast, please let me know if you can do it or i can code blindly and ask you to test @rappazzo

**Comment by rappazzo** [comment]
Once I have the VSCode dev container up, I need to install HACS to it?  Then, how do I point it to my copy of this repo?  (I am happy to discuss on discord/etc rather than hijack the github issue if there is such a place too).

**Comment by rappazzo** [comment]
https://github.com/fuatakgun/kia_uvo/compare/master...rappazzo:hyundai-usa-location?expand=1
Still not sure how to test this.  I could pull it into my HA instance I guess.

**Comment by fuatakgun** [comment]
I can see your pull request and it is a good starting point and I see that it requires PIN based call, which consumes daily API call limit of the user. I am not sure if we should place this call part of `update` or `force_update`.
- `update` is responsible to receive cached data
- `force_update` is using PIN and forcing car to refresh the data on cloud and get refreshed data from `update`

@cdnninja , I remember you had some issues with limits, what do you think?

**Comment by cdnninja** [comment]
What I did in CA was calling get location in get_status but only when odometer changed.  It also passes the pauth to reuse it i believe.  A little hacky if you look at it but works.

**Comment by fuatakgun** [comment]
did you call `get_location` inside `update` or `force_update` with odometer condition?

**Comment by AndrewDaws** [comment]
I can confirm I am also having the same Hyundai USA location truncation issue.
![Screenshot from 2021-12-02 13-45-28](https://user-images.githubusercontent.com/5349812/144484074-6a6f7970-568f-484d-8d34-765122cefb5a.png)
As a result my vehicle is showing waaaaay off where it actually is.
![Screenshot from 2021-12-02 13-46-14](https://user-images.githubusercontent.com/5349812/144484101-33c55a4e-cf67-42d2-aeac-2d7419f1e3cf.png)

**Comment by rappazzo** [comment]
I didn't add any triggers as of yet.  When I had the node red bluelinky setup, I also needed to make a separate call to get the odometer reading, so @cdnninja's solution may not work for Hyundai USA.  I don't see the odometer among the entity listing for this integration either, so that would be consistent.  

That node red setup was on a 1 hour cadence, making 3 calls - status, location, and odometer.  As far as the daily API limit, I don't think that I ever reached it.

Would it make sense to put these on a separate schedule?  Is that an option?

**Comment by fuatakgun** [comment]
what about last_updated in get_vehicle response? if data is updated on cloud response, we can trigger a PIN based call to fetch location.

**Comment by rappazzo** [comment]
Judging by the logs, this may work.  Over the last day or so in my log with DEBUG level set, it doesn't update too often.  

One thing that I am a little unclear on.  This code seems to want to format the date:
```
        vehicle_status["vehicleStatus"]["dateTime"] = (
            vehicle_status["vehicleStatus"]["dateTime"]
            .replace("-", "")
            .replace("T", "")
            .replace(":", "")
            .replace("Z", "")
        )
```
But the log looks like that isn't working:
> `'dateTime': '2021-12-01T23:05:45Z'`
Or is this log from before the alteration?

**Comment by cdnninja** [comment]
Logs are coming from line 148 so before the datetime change.  The home assistant data entity is after the changes occur.  I think your updated fork above still needs data mapping to get the value returned to Home assistant?  

As to odometer that is another bug that is open that needs to be fixed.   In Hyundai USA it is in a separate call, it is the only region like this.

**Comment by rappazzo** [comment]
I've updated my branch to be based on your odometer patch branch (with my fix).  I (mostly) copied the "if odometer changed" code from Kia CA for the updates.  Once the odometer code is merged, I will test this on my HA instance, and submit a pull request.

**Comment by cdnninja** [comment]
Good FYI on rates: https://github.com/Hacksore/bluelinky/wiki/API-Rate-Limits

Once this is working we can see if you hit the limit and how the integration handles it.  I find I don't using this logic, but we don't drive often.  In theory location is only hit if you moved between scans.  A full day of driving may max out location calls.

**Comment by rappazzo** [comment]
In the logs there are a lot more requests than I would have expected.  
```
2021-12-03 11:41:00 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration kia_uvo which has not been tested by Home Assistant
2021-12-03 11:41:07 DEBUG (MainThread) [custom_components.kia_uvo] kia_uvo - async_setup_entry started 
2021-12-03 11:41:09 DEBUG (SyncWorker_5) Get Vehicles Response
2021-12-03 11:41:18 DEBUG (SyncWorker_2) Get Vehicles Response
2021-12-03 12:11:19 DEBUG (SyncWorker_2) Get Vehicles Response
2021-12-03 12:11:21 DEBUG (SyncWorker_0) Get Vehicles Response
2021-12-03 12:11:21 DEBUG (SyncWorker_4) Get Vehicles Response
2021-12-03 12:11:23 DEBUG (SyncWorker_3) Get Vehicles Response
2021-12-03 12:11:24 DEBUG (SyncWorker_5) Get Vehicles Response
2021-12-03 12:41:25 DEBUG (SyncWorker_7) Get Vehicles Response
2021-12-03 12:41:27 DEBUG (SyncWorker_7) Get Vehicles Response
2021-12-03 12:41:27 DEBUG (SyncWorker_9) Get Vehicles Response
2021-12-03 12:41:29 DEBUG (SyncWorker_4) Get Vehicles Response
2021-12-03 12:41:29 DEBUG (SyncWorker_0) Get Vehicles Response
2021-12-03 13:11:31 DEBUG (SyncWorker_9) Get Vehicles Response
2021-12-03 13:11:33 DEBUG (SyncWorker_4) Get Vehicles Response
2021-12-03 13:11:33 DEBUG (SyncWorker_3) Get Vehicles Response
2021-12-03 13:11:35 DEBUG (SyncWorker_7) Get Vehicles Response
2021-12-03 13:11:35 DEBUG (SyncWorker_9) Get Vehicles Response
2021-12-03 13:41:37 DEBUG (SyncWorker_9) Get Vehicles Response
2021-12-03 13:41:39 DEBUG (SyncWorker_2) Get Vehicles Response
2021-12-03 13:41:39 DEBUG (SyncWorker_7) Get Vehicles Response
2021-12-03 13:41:41 DEBUG (SyncWorker_1) Get Vehicles Response
2021-12-03 13:41:41 DEBUG (SyncWorker_3) Get Vehicles Response
2021-12-03 14:11:42 DEBUG (SyncWorker_3) Get Vehicles Response
2021-12-03 14:11:44 DEBUG (SyncWorker_4) Get Vehicles Response
2021-12-03 14:11:44 DEBUG (SyncWorker_2) Get Vehicles Response
2021-12-03 14:11:47 DEBUG (SyncWorker_7) Get Vehicles Response
2021-12-03 14:11:48 DEBUG (SyncWorker_5) Get Vehicles Response
```
I lot of different threads(?) doing the same request in clusters. (11:41 (2x), 12:11 (5x), 12:41 (5x), 13:11 (5x), 13:41 (5x), 14:11 (5x)).

Is it over subscribed?  Should we try to detect a recent request + successful response?

**Comment by cdnninja** [comment]
Well that is odd -is this detailed logging or did you just truncate it?  Typically it is configured to pull in the minutes not the seconds.

**Comment by rappazzo** [comment]
I trimmed the log down to only those calls.  I can post the whole thing if you think that would help, but I would have to edit out the PII

**Comment by cdnninja** [comment]
Is this with my PR integrated and was this happening before changes were made?  Almost looks like an infinite loop of some form.

**Comment by rappazzo** [comment]
This is running the HyundaiBlueLinkAPUSA.py from your branch.  Revision 5619c710be664b20be8aadf125e64ab69b254b04.

I wouldn't say that it is an infinite loop.  Just a loop of 5 iterations.

**Comment by cdnninja** [comment]
Fair - It almost seems something else is missing from the logs or it is a different view.  Typically other methods that call get vehicle response also have other log entries.

**Comment by rappazzo** [comment]
I totally trimmed it down.  I'll post more details in a minute, but it looks like most of those calls is preceeded by `get_cached_vehicle_status` with the same thread name.  Of course, but the code, that would make sense

**Comment by rappazzo** [comment]
Here is the full log (with all of the responses removed): [kia_uvo.log](https://github.com/fuatakgun/kia_uvo/files/7652797/kia_uvo.log)

**Comment by cdnninja** [comment]
I think some of this is caused by having to call the vehicle command for both login and updating car status.   Looks like the API also logs out so every time a scan occurs it has to login again.

**Comment by rappazzo** [comment]
So, I have been running my `get_location` code for a day now, and I got rate limited (in the location query only).  So I am implementing a time check to only check the location every hour in _addition_ to the odometer check.  I also added a log warning for when the rate limit response is given.  I'll be testing this for a few days before submitting the pull request.

**Comment by rappazzo** [comment]
I am looking into overfetching of the location data, and it seems like HA is either keeping multiple instances of the Hyundai/Kia component, or it is discarding and then recreating new instances (ie, instances are being garbage collected).  My evidence of this is that that I see in multiple log entries that it doesn't have the previous vehicle state, nor does it have a copy of the stored timestamp of my last location update.

The odometer and timestamp checks do work in short term cases, but I think we would be better served if we could actually read the entity value history from HA....do you know how to do that?

**Comment by rappazzo** [comment]
OK, so the pull request is submitted for this, but there are still duplicate calls going on from (I presume) different threads.
```
2021-12-08 12:12:10 WARNING (SyncWorker_2) [custom_components.kia_uvo.HyundaiBlueLinkAPIUSA] kia_uvo - Unable to get vehicle location: {'errorSubCode': 'HT_533', 'systemName': 'HATA', 'functionName': 'findMyCar', 'errorSubMessage': 'HATA findMyCar service failed while performing the operation FindMyCar', 'errorMessage': 'Unable to send your request because a previous request is pending. Please wait and try again later.', 'errorCode': 502, 'serviceName': 'FindMyCar'}
```
Followed shortly by 
```
2021-12-08 12:12:40 DEBUG (SyncWorker_0) [custom_components.kia_uvo.HyundaiBlueLinkAPIUSA] kia_uvo - Get Vehicle Location {'head': 46, 'coord': {'alt': 90.0, 'lon': XXX, 'type': 0, 'lat': YYY}, 'accuracy': {'pdop': 10, 'hdop': 5}, 'time': '2021-12-08T17:12:25Z', 'speed': {'unit': 1, 'value': 0}}
```
Would it make sense to use the double check idiom to execute the get_location code?  Something like this:
```
if <it is OK to call get_location on the server>:
   with self.lock:
      if <it is OK to call get_location on the server>:
        do_get_location()
```

**Comment by cdnninja** [comment]
If we have thread issues I am thinking it may be a bigger issue but this part gets over my head.    I will watch my logs to see if I see any similar trends.  lets open a new issue for it.  We good to close this one?

**Comment by rappazzo** [comment]
Yeah. Good idea.  Let's close this one and start a new one on the thread problem.
